nearswap/src/twap.rs

```diff
@@ -0,0 +1,601 @@
+use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
+use near_sdk::json_types::{U128, U64};
+use near_sdk::collections::{Vector};
+use near_sdk::{env};
+use std::convert::{TryFrom,TryInto};
+
+use std::fmt;
+use crate::*;
+use crate::constants::*;
+
+#[derive(Debug)]
+pub enum Mean {
+    M_1MIN,
+    M_5MIN,
+    M_1H,
+    M_12H,
+}
+
+#[derive(BorshSerialize, BorshDeserialize)]
+pub struct Twap {
+    // last updated index in observation array
+    current_idx: u64,
+    // bool to check if previous values are overwritten by new one
+    // i.e MAX_LENGTH of array is full and we start storing observation from `0` index
+    pivoted: bool,
+    // observation array
+    pub observations: Vector<Observation>,
+
+    // mean of last 1 minutes of readings
+    // In Tuple first value indicates mean of token1 and second indicates mean of token2
+    pub mean_1min: (u128, u128),
+    // mean of last 5 minutes of readings
+    // In Tuple first value indicates mean of token1 and second indicates mean of token2
+    pub mean_5min: (u128, u128),
+    // mean of last 1 hour of readings
+    // In Tuple first value indicates mean of token1 and second indicates mean of token2
+    pub mean_1h: (u128, u128),
+    // mean of last 12 hours of readings
+    // In Tuple first value indicates mean of token1 and second indicates mean of token2
+    pub mean_12h: (u128, u128)
+}
+
+impl Twap {
+
+    pub fn new() -> Self {
+        Self {
+            current_idx: 0,
+            pivoted: false,
+            observations: Vector::new("twap".as_bytes().to_vec()),
+            mean_1min: (0, 0),
+            mean_5min: (0, 0),
+            mean_1h: (0, 0),
+            mean_12h: (0, 0)
+        }
+    }
+
+    /**
+    Initialize the oracle array by writing the first slot.
+    Called once for the lifecycle of the observations array.
+    Parameters:
+    + `time`: The time of the oracle initialization..
+    + `price1`:  price of first token
+    + `price2`: price of second token.
+    */
+    pub fn initialize(
+        &mut self,
+        time: u64,
+        price1: u128,
+        price2: u128
+    ) -> u64
+    {
+        self.observations.push( &Observation {
+            block_timestamp: time,
+            num_of_observations: 1,
+            price1_cumulative: price1,
+            price2_cumulative: price2,
+        });
+        self.current_idx = 0;
+        return 0;
+    }
+
+    /**
+    Writes an oracle observation to the array.
+    Index represents the most recently written element.
+    Parameters:
+    + `block_timestamp`: The timestamp of the new observation.
+    + `price1`: price of first token.
+    + `price2`: price of second token.
+    + `max_length`: The maximum length of TWAP array
+    */
+    pub fn write(
```

Comment:
can we make this function private, and use `log_observation` only?

Reply:
yes